### PR TITLE
KL Stability Improvements and Bug Fixes

### PR DESCRIPTION
**Pull request by vishwakftw**
This partially addresses https://github.com/probtorch/pytorch/issues/117.

### CODE REVIEW

File: torch/distributions/kl.py [hunk 1]
```diff
@@ -163,8 +163,11 @@ def kl_divergence(p, q):
 @register_kl(Bernoulli, Bernoulli)
 def _kl_bernoulli_bernoulli(p, q):
     t1 = p.probs * (p.probs / q.probs).log()
+    t1[(p.probs == 0) | (q.probs == 1)] = 0
```
Comment (fritzo) [inline torch/distributions/kl.py]: Can you explain why the `(q.probs == 1)` is needed? And why the `p.probs == 1` term below is needed?

Comment (vishwakftw) [reply torch/distributions/kl.py]: The term `t2` is `(1 - p.probs) * ln ((1 - q.probs) / (1 - p.probs))`. If `p.probs` == 1, then the log term becomes undefined, but is reduced to 0 by `1 - p.probs`.

Regarding `q.probs == 1`, that was a mistake on my side. Sorry.

**Comment by vishwakftw** [comment]
I think the git revert caused a slight problem. Maybe, if the changes are fine here, I will close this, and create a new branch with same changes and move upstream.

**Comment by fritzo** [comment]
Looks good. Do you think it's worth implementing `Categorical` and `OneHotCategorical` in this PR using your stability trick? Or are those in a separate PR?
```py
@register_kl(Categorical, Categorical)
def _kl_categorical_categorical(p, q):
    t = p.probs * (p.logits - q.logits)
    t[q.probs == 0] = float('inf')
    t[p.probs == 0] = 0
    return t.sum(-1)

@register_kl(OneHotCategorical, OneHotCategorical)
def _kl_onehotcat_onehotcat(p, q):
    return kl_divergence(p._categorical, q._categorical)
```

**Comment by vishwakftw** [comment]
I think it would be unnecessary to create a new PR for that, so I will add it in this and update the description accordingly.

**Comment by vishwakftw** [comment]
If it is fine, can I close and send upstream?

**Comment by fritzo** [comment]
Feel free to close and send upsream.

**Pull request closed without merging**